implementations/rust/ockam/ockam_identity/src/secure_channel/decryptor.rs

```diff
@@ -1,39 +1,137 @@
 use crate::identity::IdentityError;
-use crate::secure_channel::encryptor::Encryptor;
+use crate::secure_channel::encryptor::{Encryptor, KEY_RENEWAL_INTERVAL};
 use ockam_core::compat::sync::Arc;
 use ockam_core::compat::vec::Vec;
-use ockam_core::vault::{KeyId, SymmetricVault};
+use ockam_core::vault::KeyId;
 use ockam_core::Result;
+use ockam_key_exchange_xx::XXInitializedVault;
+use tracing::warn;
 
-#[derive(Clone)]
 pub(crate) struct Decryptor {
-    key: KeyId,
-    vault: Arc<dyn SymmetricVault>,
+    current_key: KeyId,
+    current_key_nonce: u64,
+
+    previous_key: Option<KeyId>,
+    vault: Arc<dyn XXInitializedVault>,
+    nonce_tracker: NonceTracker,
+}
+
+pub(crate) struct NonceTracker {
+    nonce_bitmap: u16,
+    current_nonce: u64,
 }
 
 impl Decryptor {
     /// Restore 12-byte nonce needed for AES GCM from 8 byte that we use for noise
-    fn convert_nonce_from_small(b: &[u8]) -> Result<[u8; 12]> {
+    fn convert_nonce_from_small(b: &[u8]) -> Result<(u64, [u8; 12])> {
         let bytes: [u8; 8] = b.try_into().map_err(|_| IdentityError::InvalidNonce)?;
 
         let nonce = u64::from_be_bytes(bytes);
 
-        Ok(Encryptor::convert_nonce_from_u64(nonce).1)
+        Ok((nonce, Encryptor::convert_nonce_from_u64(nonce).1))
     }
 
-    pub async fn decrypt(&self, payload: &[u8]) -> Result<Vec<u8>> {
+    pub async fn decrypt(&mut self, payload: &[u8]) -> Result<Vec<u8>> {
         if payload.len() < 8 {
             return Err(IdentityError::InvalidNonce.into());
         }
 
-        let nonce = Self::convert_nonce_from_small(&payload[..8])?;
+        let (nonce, nonce_buffer) = Self::convert_nonce_from_small(&payload[..8])?;
+
+        self.nonce_tracker.mark(nonce)?;
+
+        if nonce >= self.current_key_nonce + KEY_RENEWAL_INTERVAL {
```

Comment:
I think the sequence should be the following:
1. Obtain the key that should be used to decrypt the message (previous, current, or derive a new one)
2. Try to decrypt the message
  - If decryption fails - do nothing with the state of the Decryptor. Destroy derived keys if there were any
  - If succeeds - replace old keys with the new keys if needed, update nonce&NonceTracker

What do you think?

Reply:
This makes sense, it would protect against connection disruption attack, granted we can validate the decryption.
Basically, making the decryption an atomic operation.